olive/strategy/search_strategy.py

```diff
@@ -122,7 +135,24 @@ def _next_search_group(self, init_model_id: Optional[str] = None) -> Optional[Se
         """
         Get the next search space group and initialize the search algorithm.
         """
-        # TODO: organize the state better and make execution order more flexible using a graph
+        if not (self._spaces_groups or self._pass_by_pass_sg):
```

Comment:
Could you add some comments to explain the different if cases in the logic? 

Please correct me if i understood any of the cases wrong:
138 : joint - all flows are exhausted. pass-pass-pass - all flows exhausted, current flow exhausted 
141: No input_model_id is provided for the first call to `_next_search_group`.

151: first call, initialize by getting first flow and original input model 
156. Get best model from the previous group/pass 
...